adbc_ingest is working better, but still has some COPY INTO issues after adbc 1.2 release

### What happened?

I removed all of my previous work arounds and tried sending 24 gigs across 12 parquet files which contains 370 million rows via adbc_ingest().

From the logs I can see 1666 parquet files being generated (assuming these are 10 mb in size by default) and PUT..

But ultimately it fails with errors after 5 attempts to run COPY INTO(s).

![image](https://github.com/user-attachments/assets/a09344e3-dc19-4c45-b2c4-c38d61018163)

```
  File "../miniconda3/lib/python3.9/site-packages/adbc_driver_manager/dbapi.py", line 937, in adbc_ingest
    return _blocking_call(self._stmt.execute_update, (), {}, self._stmt.cancel)
  File "adbc_driver_manager/_lib.pyx", line 1569, in adbc_driver_manager._lib._blocking_call_impl
  File "adbc_driver_manager/_lib.pyx", line 1562, in adbc_driver_manager._lib._blocking_call_impl
  File "adbc_driver_manager/_lib.pyx", line 1295, in adbc_driver_manager._lib.AdbcStatement.execute_update
  File "adbc_driver_manager/_lib.pyx", line 260, in adbc_driver_manager._lib.check_error
adbc_driver_manager.InternalError: INTERNAL: some files not loaded by COPY command, 901 files remain after 5 retries
ERRO[0619]connection.go:410 gosnowflake.(*snowflakeConn).queryContextInternal error: context canceled
ERRO[0619]connection.go:410 gosnowflake.(*snowflakeConn).queryContextInternal error: context canceled
```

I'm going try my old workarounds using adbc_ingest() with one parquet file at a time, etc..


### Stack Trace

_No response_

### How can we reproduce the bug?

_No response_

### Environment/Setup

_No response_

I think there is a bug with the context manager in the latest Snowflake Go driver.. The error I'm seeing is a disconnect while copy into is running.

I haven't been able to figure out a work around except to revert back to the adbc version 1.1.0 drivers..

Unfortunately there are other errors cropping up now with pyarrow 17 when using adbc 1.1.0 drivers..

This always happens when loading a 2nd file with a fresh connection.. The first file always works..

Downgrading to pyarrow 16.1.0 does not uninstall Arrow Go version 17, so I'm digging around trying to figure how to clean up my environment..

```
panic: cannot allocate memory

goroutine 764 [running]:
github.com/apache/arrow/go/v17/arrow/memory/mallocator.(*Mallocator).Reallocate(0xc0009e8008, 0x400000, {0x7f59765129b0?, 0x200000, 0x7f5e17d1565e?})
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/arrow/memory/mallocator/mallocator.go:94 +0xe5
github.com/apache/arrow/go/v17/arrow/memory.(*Buffer).Reserve(0xc000fe5cc0, 0x20?)
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/arrow/memory/buffer.go:124 +0x89
github.com/apache/arrow/go/v17/parquet/internal/encoding.(*BufferWriter).Reserve(0x7f5e1979c160?, 0x7f5e195433c0?)
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/parquet/internal/encoding/types.go:387 +0xfb
github.com/apache/arrow/go/v17/parquet/internal/encoding.(*BufferWriter).Write(0xc000fdd260, {0xc1b734c000, 0xdc6d7, 0x12c251})
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/parquet/internal/encoding/types.go:420 +0x6e
github.com/apache/arrow/go/v17/parquet/file.(*serializedPageWriter).WriteDataPage(0xc000fd5900, {0x7f5e197c7090, 0xc001a7a0a0})
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/parquet/file/page_writer.go:375 +0x516
github.com/apache/arrow/go/v17/parquet/file.(*bufferedPageWriter).WriteDataPage(0xc013a9c7d0?, {0x7f5e197c7090?, 0xc001a7a0a0?})
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/parquet/file/page_writer.go:459 +0x25
github.com/apache/arrow/go/v17/parquet/file.(*columnWriter).WriteDataPage(...)
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/parquet/file/column_writer.go:214
github.com/apache/arrow/go/v17/parquet/file.(*columnWriter).buildDataPageV1(0xc000fe3980, 0x13ee2fa0?, 0xc1?, 0x10142a, {0xc1b63b0000, 0x100ba8, 0x200000})
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/parquet/file/column_writer.go:345 +0x76c
github.com/apache/arrow/go/v17/parquet/file.(*columnWriter).FlushCurrentPage(0xc000fe3980)
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/parquet/file/column_writer.go:302 +0x26f
github.com/apache/arrow/go/v17/parquet/file.(*columnWriter).commitWriteAndCheckPageLimit(0xc000fe3980, 0xc000fe3980?, 0x7f5a20706bc0?)
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/parquet/file/column_writer.go:262 +0x4f
github.com/apache/arrow/go/v17/parquet/file.(*Float64ColumnChunkWriter).WriteBatchSpaced.func1(0xc00, 0x400)
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/parquet/file/column_writer_types.gen.go:897 +0x373
github.com/apache/arrow/go/v17/parquet/file.doBatches(0x7f5e17666576?, 0x400, 0xc113ee3160)
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/parquet/file/column_writer.go:634 +0x56
github.com/apache/arrow/go/v17/parquet/file.(*Float64ColumnChunkWriter).WriteBatchSpaced(0x7f5e17658cc5?, {0x7f5a20700bc0?, 0x50?, 0x48?}, {0xc1a8070000?, 0x1005e17660ee5?, 0x7f5d5866f278?}, {0x0, 0x0, 0x0}, ...)
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/parquet/file/column_writer_types.gen.go:883 +0x105
github.com/apache/arrow/go/v17/parquet/pqarrow.writeDenseArrow(0xc013a9c230, {0x7f5e197e1ca8?, 0xc000fe3980}, {0x7f5e197db0f8, 0xc000e3f8c0?}, {0xc1a8070000, 0x20000, 0x100000}, {0x0, 0x0, ...}, ...)
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/parquet/pqarrow/encode_arrow.go:472 +0x2145
github.com/apache/arrow/go/v17/parquet/pqarrow.WriteArrowToColumn({0x7f5e197b8480, 0xc001888120}, {0x7f5e197e1ca8, 0xc000fe3980}, {0x7f5e197db0f8?, 0xc000e3f8c0?}, {0xc1a8070000, 0x20000, 0x100000}, {0x0, ...}, ...)
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/parquet/pqarrow/encode_arrow.go:232 +0x32f
github.com/apache/arrow/go/v17/parquet/pqarrow.(*arrowColumnWriter).Write(0xc113ee3b08, {0x7f5e197b8480, 0xc001888120})
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/parquet/pqarrow/encode_arrow.go:193 +0x4f3
github.com/apache/arrow/go/v17/parquet/pqarrow.(*FileWriter).WriteColumnChunked(0xc0018b4000, 0x7f5e1a9b5020?, 0xc113ee3bd8?, 0x1?)
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/parquet/pqarrow/file_writer.go:330 +0xe5
github.com/apache/arrow/go/v17/parquet/pqarrow.(*FileWriter).WriteColumnData(0xc000dc2380?, {0x7f5e197db0f8, 0xc0011eedc0})
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/parquet/pqarrow/file_writer.go:339 +0xc9
github.com/apache/arrow/go/v17/parquet/pqarrow.(*FileWriter).WriteBuffered(0xc0018b4000, {0x7f5e197d8ee0, 0xc001066360})
        /root/go/pkg/mod/github.com/apache/arrow/go/v17@v17.0.0-20240626234237-6680dcfbef42/parquet/pqarrow/file_writer.go:203 +0x3e5
github.com/apache/arrow-adbc/go/adbc/driver/snowflake.writeParquet(0x2fc6c655b1344a00?, {0x7f5e1979aa88?, 0xc0019a42a0}, 0x74852a1848ec0d8c?, 0xa00000, 0xad8b19d8d12c4235?, {{0x7f5e197ae3e0, 0xc0009e8008}, 0x0, 0x0, ...})
        /adbc/go/adbc/driver/snowflake/bulk_ingestion.go:345 +0x1cb
github.com/apache/arrow-adbc/go/adbc/driver/snowflake.runParallelParquetWriters.func3()
        /adbc/go/adbc/driver/snowflake/bulk_ingestion.go:409 +0x185
golang.org/x/sync/errgroup.(*Group).Go.func1()
        /root/go/pkg/mod/golang.org/x/sync@v0.7.0/errgroup/errgroup.go:78 +0x56
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 638
        /root/go/pkg/mod/golang.org/x/sync@v0.7.0/errgroup/errgroup.go:75 +0x96
Aborted (core dumped)
```

--- a/go/adbc/driver/snowflake/bulk_ingestion.go
+++ b/go/adbc/driver/snowflake/bulk_ingestion.go
@@ -22,7 +22,6 @@ import (
 	"bytes"
 	"compress/flate"
 	"context"
-	"database/sql"
 	"database/sql/driver"
 	"errors"
 	"fmt"
@@ -31,6 +30,7 @@ import (
 	"path"
 	"runtime"
 	"slices"
+	"strconv"
 	"strings"
 	"sync"
 	"time"
@@ -140,7 +140,7 @@ func (st *statement) ingestRecord(ctx context.Context) (nrows int64, err error) 
 	)
 
 	// Check final row count of target table to get definitive rows affected
-	initialRows, err = countRowsInTable(ctx, st.cnxn.sqldb, target)
+	initialRows, err = countRowsInTable(ctx, st.cnxn.cn, target)
 	if err != nil {
 		st.bound.Release()
 		return
@@ -195,7 +195,7 @@ func (st *statement) ingestRecord(ctx context.Context) (nrows int64, err error) 
 	}
 
 	// Check final row count of target table to get definitive rows affected
-	nrows, err = countRowsInTable(ctx, st.cnxn.sqldb, target)
+	nrows, err = countRowsInTable(ctx, st.cnxn.cn, target)
 	nrows = nrows - initialRows
 	return
 }
@@ -217,7 +217,7 @@ func (st *statement) ingestStream(ctx context.Context) (nrows int64, err error) 
 	)
 
 	// Check final row count of target table to get definitive rows affected
-	initialRows, err = countRowsInTable(ctx, st.cnxn.sqldb, target)
+	initialRows, err = countRowsInTable(ctx, st.cnxn.cn, target)
 	if err != nil {
 		return
 	}
@@ -225,7 +225,7 @@ func (st *statement) ingestStream(ctx context.Context) (nrows int64, err error) 
 	defer func() {
 		// Always check the resulting row count, even in the case of an error. We may have ingested part of the data.
 		ctx := context.WithoutCancel(ctx)
-		n, countErr := countRowsInTable(ctx, st.cnxn.sqldb, target)
+		n, countErr := countRowsInTable(ctx, st.cnxn.cn, target)
 		nrows = n - initialRows
 
 		// Ingestion, row-count check, or both could have failed
@@ -625,15 +625,20 @@ func runCopyTasks(ctx context.Context, cn snowflakeConn, tableName string, concu
 	return readyFn, stopFn, cancelFn
 }
 
-func countRowsInTable(ctx context.Context, db *sql.DB, tableName string) (int64, error) {
-	var nrows int64
+func countRowsInTable(ctx context.Context, db snowflakeConn, tableName string) (int64, error) {
+	rows, err := db.QueryContext(ctx, countQuery, []driver.NamedValue{{Value: tableName}})
+	if err != nil {
+		return 0, errToAdbcErr(adbc.StatusIO, err)
+	}
+	defer rows.Close()
 
-	row := db.QueryRowContext(ctx, countQuery, tableName)
-	if err := row.Scan(&nrows); err != nil {
+	dest := make([]driver.Value, 1)
+	if err := rows.Next(dest); err != nil {
 		return 0, errToAdbcErr(adbc.StatusIO, err)
 	}
 
-	return nrows, nil
+	n, err := strconv.Atoi(dest[0].(string))
+	return int64(n), err
 }
 
 // Initializes a sync.Pool of *bytes.Buffer.

--- a/go/adbc/driver/snowflake/bulk_ingestion.go
+++ b/go/adbc/driver/snowflake/bulk_ingestion.go
@@ -321,7 +321,7 @@ func readRecords(ctx context.Context, rdr array.RecordReader, out chan<- arrow.R
 		}
 	}
 
-	return nil
+	return rdr.Err()
 }
 
 func writeRecordToParquet(wr *pqarrow.FileWriter, rec arrow.Record) (int64, error) {
@@ -570,6 +570,15 @@ func runCopyTasks(ctx context.Context, cn snowflakeConn, tableName string, concu
 		close(stopCh)
 		close(readyCh)
 
+		// wait for any currently running copies to finish before we continue
+		if err := g.Wait(); err != nil {
+			return err
+		}
+
+		if filesToCopy.Len() == 0 {
+			return nil
+		}
+
 		maxRetries := 5 // maybe make configurable?
 		for attempt := 0; attempt < maxRetries+1; attempt++ {
 			if attempt > 0 {
@@ -585,7 +594,7 @@ func runCopyTasks(ctx context.Context, cn snowflakeConn, tableName string, concu
 
 			if filesToCopy.Len() == 0 {
 				// all files successfully copied
-				return g.Wait()
+				return nil
 			}
 		}
 

--- a/go/adbc/driver/snowflake/bulk_ingestion.go
+++ b/go/adbc/driver/snowflake/bulk_ingestion.go
@@ -536,6 +536,7 @@ func runCopyTasks(ctx context.Context, cn snowflakeConn, tableName string, concu
 	g, ctx := errgroup.WithContext(ctx)
 	g.SetLimit(concurrency)
 
+	done := make(chan struct{})
 	readyCh := make(chan struct{}, 1)
 	stopCh := make(chan interface{})
 
@@ -569,6 +570,8 @@ func runCopyTasks(ctx context.Context, cn snowflakeConn, tableName string, concu
 		defer cancel()
 		close(stopCh)
 		close(readyCh)
+
+		<-done
 
 		// wait for any currently running copies to finish before we continue
 		if err := g.Wait(); err != nil {
@@ -610,6 +613,7 @@ func runCopyTasks(ctx context.Context, cn snowflakeConn, tableName string, concu
 	}
 
 	go func() {
+		defer close(done)
 		for {
 
 			// Block until there is at least 1 new file available for copy, or it's time to shutdown

--- a/go/adbc/driver/snowflake/connection.go
+++ b/go/adbc/driver/snowflake/connection.go
@@ -618,11 +618,13 @@ func (c *connectionImpl) GetTableSchema(ctx context.Context, catalog *string, db
 	defer rows.Close()
 
 	var (
-		name, typ /* kind,*/, isnull, primary/*, unique*/ string
-		def, check, expr, comment, policyName, privDomain sql.NullString
-		fields                                            = []arrow.Field{}
+		name, typ, isnull, primary string
+		comment                    sql.NullString
+		fields                     = []arrow.Field{}
 	)
 
+	// columns are:
+	// name, type, kind, isnull, primary, unique, def, check, expr, comment, policyName, privDomain
 	dest := make([]driver.Value, len(rows.Columns()))
 	for {
 		if err := rows.Next(dest); err != nil {
@@ -634,16 +636,11 @@ func (c *connectionImpl) GetTableSchema(ctx context.Context, catalog *string, db
 
 		name = dest[0].(string)
 		typ = dest[1].(string)
-		// kind = dest[2].(string)
 		isnull = dest[3].(string)
-		def.Scan(dest[4])
 		primary = dest[5].(string)
-		// unique = dest[6].(string)
-		check.Scan(dest[7])
-		expr.Scan(dest[8])
-		comment.Scan(dest[9])
-		policyName.Scan(dest[10])
-		privDomain.Scan(dest[11])
+		if err := comment.Scan(dest[9]); err != nil {
+			return nil, errToAdbcErr(adbc.StatusIO, err)
+		}
 
 		f, err := descToField(name, typ, isnull, primary, comment)
 		if err != nil {

--- a/go/adbc/driver/snowflake/connection.go
+++ b/go/adbc/driver/snowflake/connection.go
@@ -22,6 +22,7 @@ import (
 	"database/sql"
 	"database/sql/driver"
 	"embed"
+	"errors"
 	"fmt"
 	"io"
 	"io/fs"
@@ -65,10 +66,9 @@ type snowflakeConn interface {
 type connectionImpl struct {
 	driverbase.ConnectionImplBase
 
-	cn    snowflakeConn
-	db    *databaseImpl
-	ctor  driver.Connector
-	sqldb *sql.DB
+	cn   snowflakeConn
+	db   *databaseImpl
+	ctor driver.Connector
 
 	activeTransaction bool
 	useHighPrecision  bool
@@ -102,8 +102,8 @@ func escapeSingleQuoteForLike(arg string) string {
 	}
 }
 
-func getQueryID(ctx context.Context, query string, driverConn any) (string, error) {
-	rows, err := driverConn.(driver.QueryerContext).QueryContext(ctx, query, nil)
+func getQueryID(ctx context.Context, query string, driverConn driver.QueryerContext) (string, error) {
+	rows, err := driverConn.QueryContext(ctx, query, nil)
 	if err != nil {
 		return "", err
 	}
@@ -126,42 +126,41 @@ func addLike(query string, pattern *string) string {
 	return query
 }
 
-func goGetQueryID(ctx context.Context, conn *sql.Conn, grp *errgroup.Group, objType string, catalog, dbSchema, tableName *string, outQueryID *string) {
+func goGetQueryID(ctx context.Context, conn driver.QueryerContext, grp *errgroup.Group, objType string, catalog, dbSchema, tableName *string, outQueryID *string) {
 	grp.Go(func() error {
-		return conn.Raw(func(driverConn any) (err error) {
-			query := "SHOW TERSE /* ADBC:getObjects */ " + objType
-			switch objType {
-			case objDatabases:
-				query = addLike(query, catalog)
+		query := "SHOW TERSE /* ADBC:getObjects */ " + objType
+		switch objType {
+		case objDatabases:
+			query = addLike(query, catalog)
+			query += " IN ACCOUNT"
+		case objSchemas:
+			query = addLike(query, dbSchema)
+
+			if catalog == nil || isWildcardStr(*catalog) {
 				query += " IN ACCOUNT"
-			case objSchemas:
-				query = addLike(query, dbSchema)
-
-				if catalog == nil || isWildcardStr(*catalog) {
-					query += " IN ACCOUNT"
-				} else {
-					query += " IN DATABASE " + quoteTblName(*catalog)
-				}
-			case objViews, objTables, objObjects:
-				query = addLike(query, tableName)
-
-				if catalog == nil || isWildcardStr(*catalog) {
-					query += " IN ACCOUNT"
-				} else {
-					escapedCatalog := quoteTblName(*catalog)
-					if dbSchema == nil || isWildcardStr(*dbSchema) {
-						query += " IN DATABASE " + escapedCatalog
-					} else {
-						query += " IN SCHEMA " + escapedCatalog + "." + quoteTblName(*dbSchema)
-					}
-				}
-			default:
-				return fmt.Errorf("unimplemented object type")
+			} else {
+				query += " IN DATABASE " + quoteTblName(*catalog)
 			}
+		case objViews, objTables, objObjects:
+			query = addLike(query, tableName)
 
-			*outQueryID, err = getQueryID(ctx, query, driverConn)
-			return
-		})
+			if catalog == nil || isWildcardStr(*catalog) {
+				query += " IN ACCOUNT"
+			} else {
+				escapedCatalog := quoteTblName(*catalog)
+				if dbSchema == nil || isWildcardStr(*dbSchema) {
+					query += " IN DATABASE " + escapedCatalog
+				} else {
+					query += " IN SCHEMA " + escapedCatalog + "." + quoteTblName(*dbSchema)
+				}
+			}
+		default:
+			return fmt.Errorf("unimplemented object type")
+		}
+
+		var err error
+		*outQueryID, err = getQueryID(ctx, query, conn)
+		return err
 	})
 }
 
@@ -175,12 +174,7 @@ func (c *connectionImpl) GetObjects(ctx context.Context, depth adbc.ObjectDepth,
 		showSchemaQueryID, tableQueryID                     string
 	)
 
-	conn, err := c.sqldb.Conn(ctx)
-	if err != nil {
-		return nil, err
-	}
-	defer conn.Close()
-
+	conn := c.cn
 	var hasViews, hasTables bool
 	for _, t := range tableType {
 		if strings.EqualFold("VIEW", t) {
@@ -252,25 +246,19 @@ func (c *connectionImpl) GetObjects(ctx context.Context, depth adbc.ObjectDepth,
 
 		// Detailed constraint info not available in information_schema
 		// Need to dispatch SHOW queries and use conn.Raw to extract the queryID for reuse in GetObjects query
-		gQueryIDs.Go(func() error {
-			return conn.Raw(func(driverConn any) (err error) {
-				pkQueryID, err = getQueryID(gQueryIDsCtx, "SHOW PRIMARY KEYS /* ADBC:getObjectsTables */"+suffix, driverConn)
-				return err
-			})
+		gQueryIDs.Go(func() (err error) {
+			pkQueryID, err = getQueryID(gQueryIDsCtx, "SHOW PRIMARY KEYS /* ADBC:getObjectsTables */"+suffix, conn)
+			return err
 		})
 
-		gQueryIDs.Go(func() error {
-			return conn.Raw(func(driverConn any) (err error) {
-				fkQueryID, err = getQueryID(gQueryIDsCtx, "SHOW IMPORTED KEYS /* ADBC:getObjectsTables */"+suffix, driverConn)
-				return err
-			})
+		gQueryIDs.Go(func() (err error) {
+			fkQueryID, err = getQueryID(gQueryIDsCtx, "SHOW IMPORTED KEYS /* ADBC:getObjectsTables */"+suffix, conn)
+			return err
 		})
 
-		gQueryIDs.Go(func() error {
-			return conn.Raw(func(driverConn any) (err error) {
-				uniqueQueryID, err = getQueryID(gQueryIDsCtx, "SHOW UNIQUE KEYS /* ADBC:getObjectsTables */"+suffix, driverConn)
-				return err
-			})
+		gQueryIDs.Go(func() (err error) {
+			uniqueQueryID, err = getQueryID(gQueryIDsCtx, "SHOW UNIQUE KEYS /* ADBC:getObjectsTables */"+suffix, conn)
+			return err
 		})
 
 		goGetQueryID(gQueryIDsCtx, conn, gQueryIDs, objDatabases,
@@ -300,7 +288,7 @@ func (c *connectionImpl) GetObjects(ctx context.Context, depth adbc.ObjectDepth,
 		return nil, err
 	}
 
-	args := []any{
+	args := []sql.NamedArg{
 		// Optional filter patterns
 		driverbase.PatternToNamedArg("CATALOG", catalog),
 		driverbase.PatternToNamedArg("DB_SCHEMA", dbSchema),
@@ -317,32 +305,17 @@ func (c *connectionImpl) GetObjects(ctx context.Context, depth adbc.ObjectDepth,
 		sql.Named("SHOW_TABLE_QUERY_ID", tableQueryID),
 	}
 
-	// currently only the Columns / all case still requires a current database/schema
-	// to be propagated. The rest of the cases all solely use SHOW queries for the metadata
-	// just as done by the snowflake JDBC driver. In those cases we don't need to propagate
-	// the current session database/schema.
-	if depth == adbc.ObjectDepthColumns || depth == adbc.ObjectDepthAll {
-		dbname, err := c.GetCurrentCatalog()
-		if err != nil {
-			return nil, errToAdbcErr(adbc.StatusIO, err)
-		}
-
-		schemaname, err := c.GetCurrentDbSchema()
-		if err != nil {
-			return nil, errToAdbcErr(adbc.StatusIO, err)
-		}
-
-		// the connection that is used is not the same connection context where the database may have been set
-		// if the caller called SetCurrentCatalog() so need to ensure the database context is appropriate
-		multiCtx, _ := gosnowflake.WithMultiStatement(ctx, 2)
-		_, err = conn.ExecContext(multiCtx, fmt.Sprintf("USE DATABASE %s; USE SCHEMA %s;", quoteTblName(dbname), quoteTblName(schemaname)))
-		if err != nil {
-			return nil, errToAdbcErr(adbc.StatusIO, err)
+	nvargs := make([]driver.NamedValue, len(args))
+	for i, arg := range args {
+		nvargs[i] = driver.NamedValue{
+			Name:    arg.Name,
+			Ordinal: i + 1,
+			Value:   arg.Value,
 		}
 	}
 
 	query := string(queryBytes)
-	rows, err := conn.QueryContext(ctx, query, args...)
+	rows, err := conn.QueryContext(ctx, query, nvargs)
 	if err != nil {
 		return nil, errToAdbcErr(adbc.StatusIO, err)
 	}
@@ -353,9 +326,18 @@ func (c *connectionImpl) GetObjects(ctx context.Context, depth adbc.ObjectDepth,
 
 	go func() {
 		defer close(catalogCh)
-		for rows.Next() {
+		dest := make([]driver.Value, len(rows.Columns()))
+		for {
+			if err := rows.Next(dest); err != nil {
+				if errors.Is(err, io.EOF) {
+					return
+				}
+				errCh <- errToAdbcErr(adbc.StatusInvalidData, err)
+				return
+			}
+
 			var getObjectsCatalog driverbase.GetObjectsInfo
-			if err := rows.Scan(&getObjectsCatalog); err != nil {
+			if err := getObjectsCatalog.Scan(dest[0]); err != nil {
 				errCh <- errToAdbcErr(adbc.StatusInvalidData, err)
 				return
 			}
@@ -629,24 +611,39 @@ func (c *connectionImpl) GetTableSchema(ctx context.Context, catalog *string, db
 	tblParts = append(tblParts, quoteTblName(tableName))
 	fullyQualifiedTable := strings.Join(tblParts, ".")
 
-	rows, err := c.sqldb.QueryContext(ctx, `DESC TABLE `+fullyQualifiedTable)
+	rows, err := c.cn.QueryContext(ctx, `DESC TABLE `+fullyQualifiedTable, nil)
 	if err != nil {
 		return nil, errToAdbcErr(adbc.StatusIO, err)
 	}
 	defer rows.Close()
 
 	var (
-		name, typ, kind, isnull, primary, unique          string
+		name, typ /* kind,*/, isnull, primary/*, unique*/ string
 		def, check, expr, comment, policyName, privDomain sql.NullString
 		fields                                            = []arrow.Field{}
 	)
 
-	for rows.Next() {
-		err := rows.Scan(&name, &typ, &kind, &isnull, &def, &primary, &unique,
-			&check, &expr, &comment, &policyName, &privDomain)
-		if err != nil {
+	dest := make([]driver.Value, len(rows.Columns()))
+	for {
+		if err := rows.Next(dest); err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
 			return nil, errToAdbcErr(adbc.StatusIO, err)
 		}
+
+		name = dest[0].(string)
+		typ = dest[1].(string)
+		// kind = dest[2].(string)
+		isnull = dest[3].(string)
+		def.Scan(dest[4])
+		primary = dest[5].(string)
+		// unique = dest[6].(string)
+		check.Scan(dest[7])
+		expr.Scan(dest[8])
+		comment.Scan(dest[9])
+		policyName.Scan(dest[10])
+		privDomain.Scan(dest[11])
 
 		f, err := descToField(name, typ, isnull, primary, comment)
 		if err != nil {
@@ -702,14 +699,9 @@ func (c *connectionImpl) NewStatement() (adbc.Statement, error) {
 
 // Close closes this connection and releases any associated resources.
 func (c *connectionImpl) Close() error {
-	if c.sqldb == nil || c.cn == nil {
+	if c.cn == nil {
 		return adbc.Error{Code: adbc.StatusInvalidState}
 	}
-
-	if err := c.sqldb.Close(); err != nil {
-		return err
-	}
-	c.sqldb = nil
 
 	defer func() {
 		c.cn = nil

--- a/go/adbc/driver/snowflake/snowflake_database.go
+++ b/go/adbc/driver/snowflake/snowflake_database.go
@@ -21,7 +21,6 @@ import (
 	"context"
 	"crypto/rsa"
 	"crypto/x509"
-	"database/sql"
 	"encoding/pem"
 	"errors"
 	"fmt"
@@ -452,7 +451,6 @@ func (d *databaseImpl) Open(ctx context.Context) (adbc.Connection, error) {
 	conn := &connectionImpl{
 		cn: cn.(snowflakeConn),
 		db: d, ctor: connector,
-		sqldb: sql.OpenDB(connector),
 		// default enable high precision
 		// SetOption(OptionUseHighPrecision, adbc.OptionValueDisabled) to
 		// get Int64/Float64 instead

